src/exo/cursors.py

```diff
@@ -0,0 +1,591 @@
+                raise InvalidCursorError('block was partially destroyed')
+
+            rng = range(
+                del_range.start if rng.start in del_range else rng.start,
+                del_range.stop if rng.stop in del_range[1:] else rng.stop
+            )
+
+            start = rng.start + n_diff * (rng.start >= del_range.stop)
+            stop = rng.stop + n_diff * (rng.stop >= del_range.stop)
+
+            if start >= stop:
+                raise InvalidCursorError('block no longer exists')
+
+            return range(start, stop)
+
+        return self._make_forward(new_proc, fwd_node, fwd_gap, fwd_sel)
+
+    def _delete(self):
+        """
+        This is an UNSAFE internal function for deleting a block in an AST and
+        providing a forwarding function as collateral. It is meant to be
+        package-private, not class-private, so it may be called from other
+        internal classes and modules, but not from end-user code.
+        """
+        pass_stmt = [LoopIR.Pass(None, self.parent()._node().srcinfo)]
+        return self._replace([], empty_default=pass_stmt)
+
+
+@dataclass
+class Node(Cursor):
+    _path: list[tuple[str, Optional[int]]]
+
+    # ------------------------------------------------------------------------ #
+    # Validating accessors
+    # ------------------------------------------------------------------------ #
+
+    def _node(self):
+        """
+        Gets the raw underlying node that's pointed-to. This is meant to be
+        compiler-internal, not class-private, so other parts of the compiler
+        may call this, while users should not.
+        """
+        if (n := self._node_ref()) is None:
+            raise InvalidCursorError('underlying node was destroyed')
+        return n
+
+    @cached_property
+    def _node_ref(self):
+        n = self.proc().INTERNAL_proc()
+        n = self._walk_path(n, self._path)
+        return weakref.ref(n)
+
+    # ------------------------------------------------------------------------ #
+    # Navigation (implementation)
+    # ------------------------------------------------------------------------ #
+
+    def parent(self) -> Node:
+        if not self._path:
+            raise InvalidCursorError('cursor does not have a parent')
+        return Node(self._proc, self._path[:-1])
+
+    def before(self, dist=1) -> Gap:
+        return self._translate(Gap, self._path, 1 - dist)
+
+    def after(self, dist=1) -> Gap:
+        return self._translate(Gap, self._path, dist)
+
+    def prev(self, dist=1) -> Node:
+        return self._translate(Node, self._path, -dist)
+
+    def next(self, dist=1) -> Node:
+        return self._translate(Node, self._path, dist)
+
+    # ------------------------------------------------------------------------ #
+    # Navigation (children)
+    # ------------------------------------------------------------------------ #
+
+    def child(self, attr, i=None) -> Node:
+        _node = getattr(self._node(), attr)
```

Comment:
If `child` is publicly visible, we should at least throw some kind of error (maybe not a cursor invalidation error?) when `getattr` fails?  Or I guess we're simply ok with the built-in error from `getattr`?

Reply:
I'm making `_child` private.